lib/pbench/test/unit/server/test_cache_manager.py

```diff
@@ -1552,6 +1553,152 @@ def read(self, size: int = -1) -> bytes:
+        assert cm.get_contents(md5, "", root) == {
+            "name": "",
+            "type": "DIRECTORY",
+            "uri": f"{root}/contents/",
+            "directories": [],
+            "files": [
+                {
+                    "name": "metadata.log",
+                    "size": 28,
+                    "type": "FILE",
+                    "uri": f"{root}/inventory/metadata.log",
+                }
+            ],
+        }
+
+        assert cm.get_contents(md5, "metadata.log", root) == {
+            "name": "metadata.log",
+            "type": "FILE",
+            "uri": f"{root}/inventory/metadata.log",
+            "size": 28,
+        }
+
+        with pytest.raises(CacheExtractError):
+            cm.get_contents(md5, "../..", root)
+
+        # Now that we have an unpacked directory, manually "enhance" it to test
+        # more paths.
+
+        base: Path = cm.find_dataset(md5).unpacked
+        subdir = base / "subdir"
+        subdir.mkdir()
+        ldir = base / "dir_link"
+        ldir.symlink_to(subdir)
+        lfile = base / "file_link"
+        lfile.symlink_to(Path(base / "metadata.log"))
+        badlink = base / "bad_link"
+        badlink.symlink_to("/etc/passwd")
+        fifo = base / "fifo"
+        os.mkfifo(fifo)
+        fifolink = base / "fifo_link"
+        fifolink.symlink_to(fifo)
+        assert cm.get_contents(md5, "", root) == {
+            "name": "",
+            "type": "DIRECTORY",
+            "directories": [
+                {
+                    "name": "dir_link",
+                    "type": "SYMLINK",
+                    "link": "subdir",
+                    "link_type": "DIRECTORY",
+                    "uri": f"{root}/contents/subdir",
+                },
+                {
+                    "name": "subdir",
+                    "type": "DIRECTORY",
+                    "uri": f"{root}/contents/subdir",
+                },
+            ],
+            "files": [
+                {
+                    "name": "bad_link",
+                    "type": "SYMLINK",
+                    "link": "/etc/passwd",
+                    "link_type": "BROKEN",
+                    "uri": f"{root}/inventory/bad_link",
+                },
+                {
+                    "name": "fifo",
+                    "type": "OTHER",
+                    "uri": f"{root}/inventory/fifo",
+                },
+                {
+                    "name": "fifo_link",
+                    "type": "SYMLINK",
+                    "link": "fifo",
+                    "link_type": "OTHER",
+                    "uri": f"{root}/inventory/fifo_link",
+                },
+                {
+                    "name": "file_link",
+                    "type": "SYMLINK",
+                    "link": "metadata.log",
+                    "link_type": "FILE",
+                    "uri": f"{root}/inventory/metadata.log",
+                },
```

Comment:
Any interest in a symlink to a symlink?  (Probably best not, because then we would have to decide whether the target symlink is linked to a directory, a file, broken, etc. .... 😉)

Reply:
Actually, since we use `resolve`, intermediate links should be completely transparent to us. I'm unconvinced that it'd be worth the effort of writing our own `readlink` loop. In theory, there's a hole here if we were to get a symlink *outside* the cache directory to a symlink back *into* the cache directory. One could argue it's technically illegal, but we'd never know. Since it's a controlled server environment it's also unlikely enough I find it hard to care much.

Still ... that's an interesting test case. (Not the mad dash out and back in, but an indirect link ...)

Reply:
I don't think I care about the jump-out-jump-in case:  for starters, we end up back inside the unpacked tarball, so presumably there is no "leak"; beyond that, it seems unlikely in the extreme that the jump-in symlink would ever exist (and, if it does, we have bigger problems!).